### PR TITLE
Implement Rewards/Dividends

### PR DESCRIPTION
**Pull request by blondfrogs**
- Built by @maldon
- Add functionality for dividends and reward payments

Current workflow:
- user requests an asset snapshot at a specified block height (in the future):
`requestsnapshot asset_name block_height`
- user distributes reward
`distributereward asset_name snapshot_height distribution_asset_name gross_distribution_amount ( exception_addresses )`

Update for #600, which had merge conflicts

### CODE REVIEW

**Comment by cfox** [comment]
Perhaps `getsnapshotrequest` should be replaced with `listsnapshotrequests`?  `get` doesn't seem super useful.  You're currently required to independently track your snapshot requests so you can later call `distributereward`.  `listsnapshotrequests` would show you what was there so you could then distribute or cancel.

File: src/rpc/rewards.cpp [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+
+            return responseObj;
+        }
+    }
+    else {
+        LogPrint(BCLog::REWARDS, "Failed to retrieve ownership snapshot for '%s' at height %d!\n",
+            asset_name.c_str(), snapshot_height);                
+    }
+
+    throw JSONRPCError(RPC_MISC_ERROR, std::string("Failed to distribute reward"));
```
Comment (cfox) [inline src/rpc/rewards.cpp]: This generic fallthrough message doesn't help the caller much if e.g. we failed to retrieve the ownership snapshot.

File: src/rpc/rewards.cpp [hunk 2]
```diff
@@ -0,0 +1,762 @@
...
+                "      total_amount: (number),\n"
+                "      fee: (number),\n"
+                "      expected_count: (number),\n"
+                "      actual_count: (number),\n"
+                "    }\n"
+                "  ]\n"
+                "}\n"
+
+                "\nExamples:\n"
+                + HelpExampleCli("requestsnapshot", "\"TRONCO\" 12345 \"RVN\" 1000")
```
Comment (cfox) [inline src/rpc/rewards.cpp]:
```suggestion
                + HelpExampleCli("distributereward", "\"TRONCO\" 12345 \"RVN\" 1000")
```

File: src/rpc/rewards.cpp [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+                + HelpExampleCli("requestsnapshot", "\"PHATSTACKS\" 12345 \"DIVIDENDS\" 1000 \"mwN7xC3yomYdvJuVXkVC7ymY9wNBjWNduD,n4Rf18edydDaRBh7t6gHUbuByLbWEoWUTg\"")
```
Comment (cfox) [inline src/rpc/rewards.cpp]:
```suggestion
                + HelpExampleCli("distributereward", "\"PHATSTACKS\" 12345 \"DIVIDENDS\" 1000 \"mwN7xC3yomYdvJuVXkVC7ymY9wNBjWNduD,n4Rf18edydDaRBh7t6gHUbuByLbWEoWUTg\"")
```

File: src/rpc/rewards.cpp [hunk 4]
```diff
@@ -0,0 +1,762 @@
...
+                + HelpExampleRpc("requestsnapshot", "\"TRONCO\" 34987 \"DIVIDENDS\" 100000")
```
Comment (cfox) [inline src/rpc/rewards.cpp]:
```suggestion
                + HelpExampleRpc("distributereward", "\"TRONCO\" 34987 \"DIVIDENDS\" 100000")
```

File: src/rpc/rewards.cpp [hunk 5]
```diff
@@ -0,0 +1,762 @@
...
+                + HelpExampleRpc("requestsnapshot", "\"PHATSTACKS\" 34987 \"RVN\" 100000 \"mwN7xC3yomYdvJuVXkVC7ymY9wNBjWNduD,n4Rf18edydDaRBh7t6gHUbuByLbWEoWUTg\"")
```
Comment (cfox) [inline src/rpc/rewards.cpp]:
```suggestion
                + HelpExampleRpc("distributereward", "\"PHATSTACKS\" 34987 \"RVN\" 100000 \"mwN7xC3yomYdvJuVXkVC7ymY9wNBjWNduD,n4Rf18edydDaRBh7t6gHUbuByLbWEoWUTg\"")
```

**Comment by cfox** [comment]
`getsnapshot` command doesn't show up in help in the `== Rewards ==` section!

**Comment by cfox** [comment]
Leaving this note here after talking with @blondfrogs -- We might want to add a change address to `distributereward` and make more of a best effort to divide up all the rewards regardless of respective units and quantities, sending the rest to the change address.  Might also want a "dryrun" flag where we can examine the output without creating the transactions!

File: src/rpc/rewards.cpp (hunk 1 above)
Comment (blondfrogs) [reply src/rpc/rewards.cpp]: Should be fixed with f825dcda5198fee39f80ff3b101a44bdcdd4a424

**Comment by blondfrogs** [comment]
replaced by #689

**Pull request closed without merging**